server/src/main/java/org/opensearch/search/optimization/filterrewrite/AggregatorBridge.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.optimization.filterrewrite;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PointValues;
+import org.opensearch.index.mapper.MappedFieldType;
+
+import java.io.IOException;
+import java.util.function.BiConsumer;
+
+/**
+ * This interface provides a bridge between an aggregator and the optimization context, allowing
+ * the aggregator to provide data and optimize the aggregation process.
+ *
+ * <p>The main purpose of this interface is to encapsulate the aggregator-specific optimization
+ * logic and provide access to the data in Aggregator that is required for optimization, while keeping the optimization
+ * business logic separate from the aggregator implementation.
+ *
+ * <p>To use this interface to optimize an aggregator, you should subclass this interface in this package
+ * and put any specific optimization business logic in it. Then implement this subclass in the aggregator
+ * to provide data that is needed for doing the optimization
+ *
+ * @opensearch.internal
+ */
+public abstract class AggregatorBridge {
+
+    /**
+     * The optimization context associated with this aggregator bridge.
+     */
+    OptimizationContext optimizationContext;
+
+    /**
+     * The field type associated with this aggregator bridge.
+     */
+    MappedFieldType fieldType;
+
+    void setOptimizationContext(OptimizationContext context) {
```

Comment:
I think its confusing to have both these bridge components and the context with very similar API while also adding this circular dependency.  I think the context should operate using a bridge or the bridge uses the context.  See https://github.com/mch2/OpenSearch/commit/131624beacbaf35dbc5fd19f35fea2fece3ddf2c i've tried to do this.

Reply:
Thanks, really appreciate the effort here!

---

The commit you provided is similar to a previous state of this refactoring. And there was no circular dependency, no context in bridge.

Then I try to hide the the implementation details, specifically these 2 classes, `Ranges`, `DebugInfo`. To do that, I remove them from the signature of the public methods, and add context reference to set the state.

Restricting access level of class, field and method is one goal of this refactoring. Circular dependency is a way of doing this, other way is probably pass the context to bridge method when needed and I will try that out.

---

I agree that the same set of public API from context and bridge is a bit confusing. The logic in context is general for all types of aggregator, while the logic in bridge can be different for different aggregator. I think we just need clear documentation here.

Reply:
circular dependency removed in this commit

[`9040f6f` (#14464)](https://github.com/opensearch-project/OpenSearch/pull/14464/commits/9040f6f0440d3cecbd86c7c08bdcff36f2389845)